feat(docs): KEP-2442-Support JAX Training Runtime (#2643)
kubeflow/training-operator: docs/proposals/2442-jax-runtime-trainer-v2/README.md

Reviewer: We also need test plans section.

Author: you are right. i'll add it.

# KEP-2442: JAX Runtime for Trainer V2

- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
    - [Story 3](#story-3)
- [Design Details](#design-details)
- [Alternatives](#alternatives)
- [Implementation History](#implementation-history)

## Summary

> This proposal implements key components of KEP-2170, introducing the Kubeflow Training V2 API. Specifically, it focuses on creating the TrainingRuntime and ClusterTrainingRuntime for the JAX framework, built upon the Kubernetes JobSet API. These runtimes will serve as blueprints for model training (including LLMs) within cloud-native ML pipelines. This abstraction allows Data Scientists and MLOps Engineers to easily reuse standardized runtimes and launch training jobs, particularly via the SDK, without needing deep knowledge of underlying Kubernetes complexities.

Reviewer: You can write the summary section in these formats:

1. "This document outlines a proposal to support Jax Runtime in Kubeflow Trainer V2"
2. Then describe its necessity and benefits in a few sentences, like "Built upon Kubernetes JobSet API, the Jax runtime ...."
3. Other things you want to write
4. Finally, mention the main framework you use, like "Thanks to the [Kubeflow Trainer Pipeline Framework,](https://github.com/kubeflow/trainer/tree/master/docs/proposals/2170-kubeflow-trainer-v2#pipeline-framework) we can seamlessly support Jax Runtime in Kubeflow Trainer as a runtime plugin"

Author: done.

## Motivation

> JAX is a powerful Computation library created by Google, It is widely used in machine learning research and ranks as the third most wide used deep learning framework. JAX is not only a deep learning framework but suggests its potential in 

Reviewer: ```suggestion
JAX is a powerful ML framework created by Google, It is widely used in the machine learning research and ranks as the third most widely used deep learning frameworks. JAX is not only a deep learning framework but suggests its potential in 
```

> differential programming, large-scale physics simulations and many more. These usecases added on top of the new Runtime API for distributed training or calculation of objectives enables new users on top of kubeflow trainer, like distributed simulation or training of LLM prototypes developed with JAX, like vast models from Google Deep mind.

Reviewer: You can list the benefits with:

```
1.
2.
3.
```

And It might be more clear.

Author: done

In General the motivation it to enable users to use Single-Program Multi-Data (SPMD) pattern with JAX Framework. 

### Goals

- Implement ClusterTrainingRuntime for JAX, supporting multi-controller JAX
- Build the necessary Docker images for JAX worker nodes used by the runtime

> - Implement the solution to work on CPU and GPU using OpenMPI backend

Reviewer: ```suggestion
- Implement the solution to work on CPU and GPU
```

We'd better not to mention the detailed design plans in Goals section.

- Document user guides for utilizing JAX ClusterTrainingRuntimes
- Test the implementation thoroughly using unit tests and end-to-end (E2E) tests

### Non-Goals

- No TPU support (duo to lack of available TPU testing infrastructure)
- No GPU testing, tests will use CPUs

> - No Custom MLPolicy, since using OpenMPI, it can handle required parameters

Reviewer: Could you please show more reference materials for us? For example, other design plans against OpenMPI.

Or you could list these different design plans and show their pros/cons, so that we can compare them and choose the best one.

Author: oh, now i see, I'll do so.

Author: done.

> - No mechanism for finding processes, since with OpenMPI backend JAX can find all processes automatically

Reviewer: Like above, We'd better not to mention the detailed design plans in Goals section.
```suggestion
```

- Complex end-to-end examples demonstrating the runtimes (focus is on the runtime implementation itself; examples may require specific infrastructure)

> ## Proposal

Reviewer: FYR, we can describe the high-level design ideas in the Proposal section.

### User Stories

#### Story 1

As a MLOps Engineer or Platform Engineer, I want to manage JAX distributed training jobs using the Kubeflow Trainer V2, so then I can provide blueprints for training of machine learning models on a kubernetes cluster to engineering teams.

#### Story 2

As a Data Scientist, I want to use the Trainer V2 SDK to run a distributed training job from notebook, in this way I can incorporate multiple devices for my training task.

#### Story 3

> As a Research Scientist, I want to train prototype of my new LLM model coded with JAX on a distributed training setup on my company Kubernetes cluster, Kubeflow Trainer V2 with JAX ClusterTrainingRuntime will enable this for me.

Reviewer: We'd better provide detailed yaml files or python codes for these user stories.

Like: https://github.com/kubeflow/trainer/tree/master/docs/proposals/2170-kubeflow-trainer-v2#example-of-trainjob

Author: done.

In general the motivation is to enable users to use Single-Program Multi-Data (SPMD) pattern with JAX Framework, however there are other reasons like ensure backward compatibility with Trainer V1, which previously included JAX support, allowing existing users to transition smoothly while taking advantage of the enhanced Runtime API.

Finally with this design, Platform Admins can define standardized training runtimes, while AI Practitioners can easily customize them, through a simple SDK interface, without needing to understand Kubernetes internals.

**Benefits**

1. Leverage JAX for differential programming and large-scale simulations
2. Enable distributed training or objective computation using the new Runtime API
3. Support prototyping and training of large JAX-based LLMs within Kubeflow Trainer

### Goals

- Implement ClusterTrainingRuntime for JAX, supporting distributed training with JAX (e.g. multi-controller JAX)
- Build the necessary Docker images for JAX worker nodes used by the runtime

> - Implement the solution to work on CPU and GPU

Author: yes, I received 1k$ from dev-rel team of google.

> - Implement the solution to work on CPU and GPU

Author: done.

- Integrate with SDK and address any necessary enhancements
- Document user guides for utilizing JAX ClusterTrainingRuntimes
- Test the implementation thoroughly using unit tests and end-to-end (E2E) tests

### Non-Goals

- No TPU testing, tests will use CPU

> - No GPU testing, tests will use CPU

Author: you are right.

> - No GPU testing, tests will use CPU

Author: done.

## Proposal

### User Stories

#### Story 1

As a Platform Admin, I want to manage JAX distributed training jobs using the Kubeflow Trainer V2, so then I can provide blueprints for training of machine learning models on a kubernetes cluster to engineering teams.

#### Story 2

As an AI Practitioner, I want to use the Trainer V2 SDK to run a distributed training job from notebook, in this way I can incorporate multiple devices for my training task.

The Python SDK with JAXRuntime may look as follows:

```python
from kubeflow.trainer import TrainerClient, CustomTrainer

# Add logic using JAX methods

> def jax_train_mnist(args):

Author: done.

| **Pod** | A Kubernetes Pod runs a single JAX process. Scheduled on a node and may use one or more GPUs depending on resource specifications. |
| **Node** | A Kubernetes Node is a worker machine. In multi-node JAX jobs, each node typically runs one pod, mapping to one JAX host. |

### JAX Training Workflow

This section explains the architecture and flow of executing a distributed JAX training job using Kubeflow, as depicted in the diagram.

![user-roles](./drawing.drawio.svg)

| **Actor / Component**   | **Action**                                | **Details**                                                                                                                     |
| ----------------------- | ----------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- |
| Platform Admin          | Prepares the Cluster Training Runtime | Defines container image, entrypoint, framework (e.g., JAX), and resource needs. Setup reusable for training jobs.               |

> | System                  | Retrieves the Training Runtime Spec   | Fetched automatically when a user requests a training job to determine execution details.                                       |

Author: done.

| AI Practitioner         | Creates the Training Job              | Uses Kubeflow Python SDK or `kubectl`. Provides training function (e.g., `jax_train_mnist`), arguments, and node configuration. |

> | Runtime | Creates and Submits a JobSet          | Training job spec is translated into a JobSet (group of coordinated jobs).                                                      |

Author: done.

| JobSet Controller | Launches Distributed Jobs             | JobSet spawns multiple Kubernetes Jobs, each pod runs a JAX training process instance.                                          |
| Headless Service        | Connects Pods for Communication           | Enables direct pod-to-pod communication for gradient sharing and coordination.                                                  |
| Cluster (Pods)    | Executes Distributed Training         | Each pod runs JAX+Python code, collaborating to complete training across available hardware.                                    |

### Defining Distributed JAX with MLPolicy

The number of **JAX hosts** is configured using the `numNodes` field in the **MLPolicy** section of the **ClusterTrainingRuntime**. Each host runs a single JAX process inside a Pod.

#### JAXMLPolicySource

> `JAXMLPolicySource` configures JAX distributed initialization, backend, devices, and precision.

Reviewer 2: Please adjust this text to say that `JAXMLPolicySource` API indicates that JAX plugin should be activated and extension framework will set the appropriate values for JAX distributed environment, etc.

2. Enable distributed training or objective computation using the new Runtime API
3. Support prototyping and training of large JAX-based LLMs within Kubeflow Trainer

### Goals

- Implement ClusterTrainingRuntime for JAX, supporting multi-controller JAX
- Build the necessary Docker images for JAX worker nodes used by the runtime
- Implement the solution to work on CPU and GPU
- Document user guides for utilizing JAX ClusterTrainingRuntimes
- Test the implementation thoroughly using unit tests and end-to-end (E2E) tests

### Non-Goals

- No TPU support (duo to lack of available TPU testing infrastructure)
- No GPU testing, tests will use CPUs
- No Custom MLPolicy, since using OpenMPI, it can handle required parameters
- Complex end-to-end examples demonstrating the runtimes (focus is on the runtime implementation itself; examples may require specific infrastructure)

## Proposal

### User Stories

#### Story 1

As a MLOps Engineer or Platform Engineer, I want to manage JAX distributed training jobs using the Kubeflow Trainer V2, so then I can provide blueprints for training of machine learning models on a kubernetes cluster to engineering teams.

> The ClusterTrainingRuntime may look as follows:

Author: yes, it totally makes sense. i'll do so. thanks

> The ClusterTrainingRuntime may look as follows:

Author: done.

    print(f"Name: {r.name}, Framework: {r.trainer.framework.value}, Trainer Type: {r.trainer.trainer_type.value}\n")
    print(f"Entrypoint: {r.trainer.entrypoint[:3]}\n")

    if r.name == "jax-distributed":
        jax_runtime = r

# request training with jax runtime
job_id = TrainerClient().train(
    trainer=CustomTrainer(
        func=jax_train_mnist,
        func_args=args,
        num_nodes=3,
    ),
    runtime=jax_runtime,
)
```

#### Story 3

As a Research Scientist, I want to train prototype of my new LLM model coded with JAX on a distributed training setup on my company Kubernetes cluster, Kubeflow Trainer V2 with JAX ClusterTrainingRuntime will enable this for me.

## Design Details

In order to address this functionality, we propose the following design:

### Communication Backend

#### OpenMPI

**Pros:**

* Compatible with existing MPI runtime in Kubeflow Trainer v2, making deployment easier.

**Cons:**

* Typically requires more complex environment setup compared to simpler backends like Gloo.

#### Gloo

**Pros:**

* Lightweight and simple to use.

**Cons:**

* Significantly slower than OpenMPI (10–20×) for distributed JAX training on CPUs and GPUs.

> * Less optimized for multi-node scaling and lacks native support for high-speed interconnects like InfiniBand.

Reviewer 3: > asking me, i'd say, open-mpi has upper hand, and lack of gpu support in gloo is a significant design decision maker, at least for me. i'd appreciate @sandipanpanda feedback on it.

The main motive behind using gloo was it was simple to set up and use, making development iterations faster, in long term open-mpi provides more value

### Defining JAX Processes with MLPolicy

The number of JAX processes can be defined using the `numNodes` field within the `mlPolicy` section of the `ClusterTrainingRuntime` configuration. This allows for specifying how many JAX processes/controller run in the distributed setup.

> ![user-roles](./drawing.drawio.svg)

Reviewer: Could you also describe the whole procedure step-by-step with words? I think, it will be more clear to readers if the graph is companied with descriptive words:)

Author: done.

                          echo "--------------------------------------"
                          set -e
                          mpirun --version
                          python --version
                          pip list
        - name: node
          template:
            spec:
              template:
                spec:
                  containers:
                    - name: node
                      image: ghcr.io/kubeflow/trainer/jax-runtime
                      securityContext:
                        runAsUser: 1000
                      command:
                        - /usr/sbin/sshd
                      args:
                        - -De
                        - -f
                        - /home/mpiuser/.sshd_config
                      readinessProbe:
                        tcpSocket:
                          port: 2222
                        initialDelaySeconds: 5
```

### Gloo

**Pros:**

* Lightweight and simple to use.

**Cons:**

* Significantly slower than OpenMPI (10–20×) for distributed JAX training on CPUs and GPUs.
* Less optimized for multi-node scaling and lacks native support for high-speed interconnects like InfiniBand.

> ### Defining JAX Processes with MLPolicy

Author: done.

> The number of JAX hosts can be defined using the `numNodes` field within the `mlPolicy` section of the `ClusterTrainingRuntime` configuration. This allows for specifying how many JAX processes/controller run in the distributed setup.

Author: no, realistically not. since in jax all devices attached to a host/controller is managed by the only process initiated in the host/pod.

> The number of JAX hosts can be defined using the `numNodes` field within the `mlPolicy` section of the `ClusterTrainingRuntime` configuration. This allows for specifying how many JAX processes/controller run in the distributed setup.

Author: i updated design details, and to make it more clear i added a new section to address this concepts in jax setup.

> The number of JAX hosts can be defined using the `numNodes` field within the `mlPolicy` section of the `ClusterTrainingRuntime` configuration. This allows for specifying how many JAX processes/controller run in the distributed setup.

Reviewer: @mahdikhashan Thanks for this!

* This spec defines how the job should be executed across the cluster.

#### 3. Data Scientist or ML Researcher Creates a TrainJob

* A **Data Scientist / ML Research Scientist** defines a training job using either:

  * The **Kubeflow Python SDK**, or
  * A `kubectl` command.
* The training function (e.g., `jax_train_mnist`) is provided, along with any arguments and configuration (e.g., `num_nodes` for distributed training).

#### 4. TrainJob Submits a JobSet

* Once submitted, the **TrainJob** uses the runtime spec to create a **JobSet**.
* A **JobSet** is a group of Kubernetes Jobs that work together to complete the training process.

#### 5. Distributed Jobs Are Launched

* The **JobSet** schedules and launches multiple **Kubernetes Jobs**.
* Each job represents an individual **JAX training process**, running in its own pod.

#### 6. Headless Service Enables Communication

* A **Headless Kubernetes Service** is created to allow **direct communication** between the JAX pods.
* This is critical for JAX’s distributed training features, such as gradient synchronization and parallel execution.

> #### 7. JAX Processes Execute the Training

Author: yes, sure. done.

- [Proposal](#proposal)
  - [User Stories](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
    - [Story 3](#story-3)
- [Design Details](#design-details)
    - [Communication Backend](#communication-backend)
        - [OpenMPI](#openmpi)
        - [Gloo](#gloo)
    - [Defining JAX Processes with MLPolicy](#defining-jax-processes-with-mlpolicy)
    - [JAX Training Workflow Flow](#jax-training-workflow-flow)
- [Test Plan](#test-plan)
    - [End-to-End (E2E) Tests](#end-to-end-e2e-tests)
    - [Working Examples](#working-examples)
    - [Unit and Integration Tests](#unit-and-integration-tests)
- [Implementation History](#implementation-history)

## Summary

This document outlines a proposal to support the JAX Runtime in Kubeflow Trainer V2. Built upon the Kubernetes JobSet API, the JAX Runtime enables training and fine-tuning workloads using the JAX framework on Kubernetes. Instead of relying on framework-specific CRDs, Trainer V2 introduces a unified abstraction through TrainingRuntime and TrainJob. The JAX Runtime implements this abstraction to serve as a reusable blueprint for model training tasks, including large language models (LLMs).

>  With this design, Platform Engineers can define standardized training runtimes, while Data Scientists can easily customize them, through a simple SDK interface, without needing to understand Kubernetes internals. The Kubeflow Trainer Pipeline Framework further simplifies integration, allowing the JAX Runtime to be added seamlessly as a runtime plugin.

Author: yes, you are right. done.

                          set -e
                          mpirun --version
                          python --version
                          pip list
        - name: node
          template:
            spec:
              template:
                spec:
                  containers:
                    - name: node
                      image: ghcr.io/kubeflow/trainer/jax-runtime
                      securityContext:
                        runAsUser: 1000
                      command:
                        - /usr/sbin/sshd
                      args:
                        - -De
                        - -f
                        - /home/mpiuser/.sshd_config
                      readinessProbe:
                        tcpSocket:
                          port: 2222
                        initialDelaySeconds: 5
```

### Gloo

**Pros:**

* Lightweight and simple to use.

**Cons:**

* Significantly slower than OpenMPI (10–20×) for distributed JAX training on CPUs and GPUs.
* Less optimized for multi-node scaling and lacks native support for high-speed interconnects like InfiniBand.

### Defining JAX Processes with MLPolicy

> The number of JAX processes can be defined using the `numNodes` field within the `mlPolicy` section of the `ClusterTrainingRuntime` configuration. This allows for specifying how many JAX processes/controller run in the distributed setup.

Reviewer: Do you mean JAX's host or processes? Host is `jax.devices()`, and process is `jax.local_devices()`. And they correspond to `numNode` and `numProcPerNode` respectively.

Also, do we need other JAX specific parameters?

Author: um, host is the correct one.

* A **Headless Kubernetes Service** is created to allow **direct communication** between the JAX pods.
* This is critical for JAX’s distributed training features, such as gradient synchronization and parallel execution.

#### 7. JAX Processes Execute the Training

* Each pod runs the user-defined training logic using JAX and Python.
* The processes communicate and synchronize through the headless service, executing distributed training on the assigned hardware (e.g., Cloud GPUs).

## Test Plan

The testing strategy will focus on validating functionality, usability, and integration of the proposed `TrainingRuntime` mechanism for distributed training workloads. It includes the following components:

### End-to-End (E2E) Tests

* **Environment**: Deploy workloads in lightweight local Kubernetes clusters using tools like `kind` or `minikube`.
* **Workloads**: Run simple distributed training examples such as MNIST **JAX**.
* **Validation Goals**:

  * Ensure correct creation of `JobSet` resources.
  * Validate successful job execution and error handling paths.
  * Confirm compatibility with `TrainingRuntime` configurations.

### Working Examples

* Provide clear, runnable examples:

>   * **YAML manifests** showcasing the full lifecycle of a `TrainingRuntime` job.

Author: done.

These usecases added on top of the new Runtime API for distributed training or calculation of objectives enables new users on top of Kubeflow Trainer, like distributed simulation or training of LLM prototypes developed with JAX, like vast models from Google DeepMind.

In general the motivation is to enable users to use Single-Program Multi-Data (SPMD) pattern with JAX Framework.

With this design, Platform Engineers can define standardized training runtimes, while Data Scientists can easily customize them, through a simple SDK interface, without needing to understand Kubernetes internals.

**Benefits**

1. Leverage JAX for differential programming and large-scale simulations
2. Enable distributed training or objective computation using the new Runtime API
3. Support prototyping and training of large JAX-based LLMs within Kubeflow Trainer

### Goals

- Implement ClusterTrainingRuntime for JAX, supporting multi-controller JAX
- Build the necessary Docker images for JAX worker nodes used by the runtime
- Implement the solution to work on CPU and GPU
- Document user guides for utilizing JAX ClusterTrainingRuntimes

> - Test the implementation thoroughly using unit tests and end-to-end (E2E) tests

Author: sure thing. i'll mention it.

> - Test the implementation thoroughly using unit tests and end-to-end (E2E) tests

Author: done

### Non-Goals

> - No TPU support (duo to lack of available TPU testing infrastructure)

Author: I'll mention how-to's about it in the proposal, you are right.

- No GPU testing, tests will use CPUs

## Proposal

### User Stories

#### Story 1

As a Platform Engineer, I want to manage JAX distributed training jobs using the Kubeflow Trainer V2, so then I can provide blueprints for training of machine learning models on a kubernetes cluster to engineering teams.

#### Story 2

As a Data Scientist, I want to use the Trainer V2 SDK to run a distributed training job from notebook, in this way I can incorporate multiple devices for my training task.

The Python SDK with JAXRuntime may look as follows:

```python
from kubeflow.trainer import TrainerClient, CustomTrainer

def jax_train_mnist():
    # TODO: Add training logic using JAX
    pass

# Select the JAX runtime
client = TrainerClient()
jax_runtime = next(r for r in client.list_runtimes() if r.name == "jax-distributed")

# Launch training job
job_id = client.train(
    trainer=CustomTrainer(func=jax_train_mnist, func_args=args, num_nodes=3),
    runtime=jax_runtime,
)
```

> ## Design Details

Author: yes, sure. I'll add the details for multiple backend support @Electronic-Waste suggested, then I'll consider details for in a plugin. hope it is fine with you.

JAX uses **Single Program Multiple Data (SPMD)** across these processes.

**Devices**

Devices refer to individual compute units on a host (like CPU cores, GPUs, or TPUs). JAX handles device detection automatically and runs parallel computations using `jax.pmap`, `jax.shard_map`, or `pjit`.

Each JAX process has access to **all devices on its host**. There is **no need to spawn multiple processes per GPU**, unlike PyTorch.

**Pod**

A **Pod** in Kubernetes runs a JAX process. It is scheduled on a node and may use one or more GPUs depending on the resource specification.

**Node**

In Kubernetes, a **Node** is a worker machine in the cluster. When we run multi-node JAX jobs, each node typically runs one pod, which maps to one JAX host.

### JAX Training Workflow

This section explains the architecture and flow of executing a distributed JAX training job using Kubeflow, as depicted in the diagram.

> ![user-roles](./drawing.drawio.svg)

Author: sure. consider it done.

> ![user-roles](./drawing.drawio.svg)

Author: done!

- A **Data Scientist** or **ML Researcher** creates a training job using:
  - The **Kubeflow Python SDK**, or
  - A `kubectl` command.
- They provide the training function (e.g., `jax_train_mnist`), any needed arguments, and settings like how many nodes to use.

#### 4. JobSet is Created and Submitted
- The training job uses the runtime spec to create a **JobSet**, a group of jobs working together to train the model.

#### 5. Distributed Jobs Start Running
- The **JobSet** launches multiple **Kubernetes Jobs**.
- Each job runs one instance of the **JAX training process** in its own pod.

#### 6. Headless Service Connects the Jobs
- A **Headless Service** allows the pods to **communicate directly** for tasks like sharing gradients and coordinating training.

#### 7. Training Runs Across the Cluster
- Each pod runs the training code using **JAX and Python**.
- The pods work together to complete the distributed training on the available hardware.

### Defining Distributed JAX with MLPolicy

The number of **JAX hosts** is configured using the `numNodes` field in the **MLPolicy** section of the **ClusterTrainingRuntime**. Each host runs a single JAX process inside a Pod.

> ### Communication Backend

Author: yes and no (for the first part). yes, you are right (for second part), i digged deeper and will update the proposal, for some backends it needs, for some not. i'll add all required and optional env for multiple backend scenario.

**ClusterTrainingRuntime Design**

```yaml
apiVersion: trainer.kubeflow.org/v1alpha1
kind: ClusterTrainingRuntime
metadata:
  name: jax-distributed
spec:
  mlPolicy:
    numNodes: 4
  template:
    spec:
      successPolicy:
        operator: All
      replicatedJobs:
        - name: process
          template:
            spec:
              template:
                spec:
                  containers:
                    - name: node
                      image: ghcr.io/kubeflow/trainer/jax-runtime
                      securityContext:
                        runAsUser: 1000
                      command:
                        - bash
                        - -c
                        - |
                          echo "JAX Distributed Runtime"
                          echo "--------------------------------------"
                          set -e
                          python --version
                          pip list | grep jax
```

## Test Plan

The testing strategy will focus on validating functionality, usability, and integration of the proposed `TrainingRuntime` mechanism for distributed training workloads. It includes the following components:

> ### End-to-End (E2E) Tests

Author: yes, you are right. this question need to be answered.

- Test the implementation thoroughly using unit tests and end-to-end (E2E) tests

### Non-Goals

- No TPU support (duo to lack of available TPU testing infrastructure)
- No GPU testing, tests will use CPUs

## Proposal

### User Stories

#### Story 1

As a Platform Engineer, I want to manage JAX distributed training jobs using the Kubeflow Trainer V2, so then I can provide blueprints for training of machine learning models on a kubernetes cluster to engineering teams.

#### Story 2

As a Data Scientist, I want to use the Trainer V2 SDK to run a distributed training job from notebook, in this way I can incorporate multiple devices for my training task.

The Python SDK with JAXRuntime may look as follows:

```python
from kubeflow.trainer import TrainerClient, CustomTrainer

def jax_train_mnist():
    # TODO: Add training logic using JAX
    pass

# Select the JAX runtime
client = TrainerClient()
jax_runtime = next(r for r in client.list_runtimes() if r.name == "jax-distributed")

# Launch training job
job_id = client.train(
    trainer=CustomTrainer(func=jax_train_mnist, func_args=args, num_nodes=3),
    runtime=jax_runtime,
)
```

> ## Design Details

Author: all of the envs are optional, however, i'll dig deeper and update you.

- A **Data Scientist** or **ML Researcher** creates a training job using:
  - The **Kubeflow Python SDK**, or
  - A `kubectl` command.
- They provide the training function (e.g., `jax_train_mnist`), any needed arguments, and settings like how many nodes to use.

#### 4. JobSet is Created and Submitted
- The training job uses the runtime spec to create a **JobSet**, a group of jobs working together to train the model.

#### 5. Distributed Jobs Start Running
- The **JobSet** launches multiple **Kubernetes Jobs**.
- Each job runs one instance of the **JAX training process** in its own pod.

#### 6. Headless Service Connects the Jobs
- A **Headless Service** allows the pods to **communicate directly** for tasks like sharing gradients and coordinating training.

#### 7. Training Runs Across the Cluster
- Each pod runs the training code using **JAX and Python**.
- The pods work together to complete the distributed training on the available hardware.

### Defining Distributed JAX with MLPolicy

The number of **JAX hosts** is configured using the `numNodes` field in the **MLPolicy** section of the **ClusterTrainingRuntime**. Each host runs a single JAX process inside a Pod.

> ### Communication Backend

Author: to my current understanding of jax, to config for multi-processor setup, XLA will handle the multi-gpu communication setup of NCCL and other cuda related packages. that said, what i suggest is to consider this in docker image and leave the decision to the user, user can give the jax engine the list of devices in the order it prefers to use by initialization config (and user can define them in python code). so having a cuda enabled image and installing jax[cuda12] in it is enough. 

with the above argument, i'm against the idea of design for the cluster-training-runtime level config and with having this in docker image, i'd leave it to the user (platform engineer one) - they can also define their own images.

ref: https://docs.jax.dev/en/latest/installation.html#pip-installation-nvidia-gpu-cuda-installed-locally-harder

and in between gloo and openmpi, i'm again not concerned here, we can have multiple cluster-training-runtime, since there are blueprints, so as long as jax has the option to define the backend, i'd leave this decision to the end user and keep this blueprint(template) straight forward and simple.

> ### Communication Backend

Reviewer: > i'd leave it to the user (platform engineer one) - they can also define their own images.

So, do you mean that we should define the communication backend in the pod envs in replicatedJobs of ClusterTrainingRuntime?

Similarly, we can also define these configuration items in pod envs of replicatedJobs of `torch-distributed` ClusterTrainingRuntime. But for the convenience of users, we abstract some of the features of torch like `num_proc_per_node` and elastic policies:

https://github.com/kubeflow/trainer/blob/f8d71ece82c5125136917be62b719f6b12b902b2/pkg/apis/trainer/v1alpha1/trainingruntime_types.go#L173-L185

However, I'm not against your ideas. I just think that we'd better make our design user-centric. And separate configuration items, like that in `TorchMLPolicySource`, are more user-friendly for users rather than env variables in pod envs.

WDYT? @mahdikhashan @kubeflow/wg-training-leads @astefanutti

> ### Communication Backend

Reviewer 2: @mahdikhashan Please can you explain why do you want to leverage `mpirun` to start JAX job ?

> ### Communication Backend

Author: yes, you are right. these questions need to be addressed.

>In Kubernetes, a host maps to a Node, and in the runtime implementation, one Pod is typically scheduled per host.

**JAX Process/Controller**

A **JAX process** (or sometimes called a **controller**) is a Python process running the JAX program. There is exactly one JAX process per host. This process is responsible for executing the training loop, managing all local devices, and communicating with other JAX processes over the network for synchronization

JAX uses **Single Program Multiple Data (SPMD)** across these processes.

**Devices**

Devices refer to individual compute units on a host (like CPU cores, GPUs, or TPUs). JAX handles device detection automatically and runs parallel computations using `jax.pmap`, `jax.shard_map`, or `pjit`.

Each JAX process has access to **all devices on its host**. There is **no need to spawn multiple processes per GPU**, unlike PyTorch.

**Pod**

A **Pod** in Kubernetes runs a JAX process. It is scheduled on a node and may use one or more GPUs depending on the resource specification.

**Node**

> In Kubernetes, a **Node** is a worker machine in the cluster. When we run multi-node JAX jobs, each node typically runs one pod, which maps to one JAX host.

Author: sure, i'll update it.

#### 2. Training Runtime is Retrieved
- When a user requests a training job, the system **fetches the runtime spec** to know how to run the job.

#### 3. AI Practitioner Creates the Training Job
- A **AI Practitioners** creates a training job using:
  - The **Kubeflow Python SDK**, or
  - A `kubectl` command.
- They provide the training function (e.g., `jax_train_mnist`), any needed arguments, and settings like how many nodes to use.

#### 4. JobSet is Created and Submitted
- The training job uses the runtime spec to create a **JobSet**, a group of jobs working together to train the model.

#### 5. Distributed Jobs Start Running
- The **JobSet** launches multiple **Kubernetes Jobs**.
- Each job runs one instance of the **JAX training process** in its own pod.

#### 6. Headless Service Connects the Jobs
- A **Headless Service** allows the pods to **communicate directly** for tasks like sharing gradients and coordinating training.

#### 7. Training Runs Across the Cluster
- Each pod runs the training code using **JAX and Python**.
- The pods work together to complete the distributed training on the available hardware.

### Defining Distributed JAX with MLPolicy

> The number of **JAX hosts** is configured using the `numNodes` field in the **MLPolicy** section of the **ClusterTrainingRuntime**. Each host runs a single JAX process inside a Pod.

Author: yes. consider it done.

  // +kubebuilder:default="nccl"  
  // +kubebuilder:validation:Enum=nccl;gloo;mpi
  TargetBackend *string `json:"targetBackend,omitempty"`

  // Platforms is comma-separated list of platform names
  // specifying which platforms jax should initialize
  // +kubebuilder:default="gpu,tpu,cpu"
  Platforms *string `json:"platform,omitempty"`

  // Whether to disable JAX compilation optimizations.  
  // +kubebuilder:default=false
  DisableJIT *bool `json:"disableJIT,omitempty"`

  // Check for and raise errors on NaNs
  // +kubebuilder:default=false
  DebugNaNs *bool `json:"debugNaNs,omitempty"`

  // Set default precision for matrix multiplication
  // +kubebuilder:validation:Enum=default;high;highest;bfloat16;tensorfloat32;float32;
  // ANY_F8_ANY_F8_F32;ANY_F8_ANY_F8_F32_FAST_ACCUM;ANY_F8_ANY_F8_ANY;
  // ANY_F8_ANY_F8_ANY_FAST_ACCUM;F16_F16_F16;F16_F16_F32;BF16_BF16_BF16;
  // BF16_BF16_F32;BF16_BF16_F32_X3;BF16_BF16_F32_X6;BF16_BF16_F32_X9;
  // TF32_TF32_F32;TF32_TF32_F32_X3;F32_F32_F32;F64_F64_F64
  DefaultMatMulPrecision *string `json:"defaultMatmulPrecision,omitempty"`

  // Additional specific configurations.  
  // +listType=map  
  // +listMapKey=name  

>   ExtraEnv []corev1.EnvVar `json:"extraEnv,omitempty"`

Reviewer: Can we first support only some of the necessary fields, and add more if users require it? We can always have a minimum version first, and iterate over it in the future:)

WDYT? @andreyvelich @mahdikhashan

  // Distributed contains explicit args used when calling jax.distributed.initialize().
  // This should be provided when not relying on automatic cluster detection (Slurm, MPI launcher, Cloud TPU, etc).
  Distributed *DistributedConfig `json:"distributed,omitempty"`
}

type DistributedConfig struct {
  // CoordinatorAddress is the address (host:port) of the coordinator process.
  // +kubebuilder:validation:Required

>   CoordinatorAddress string `json:"coordinatorAddress"`

Reviewer: Is it possible for us to generate coordinator address like we did in PyTorch?

https://github.com/kubeflow/trainer/blob/aca63664ce60226595c83fc50bb8a39b47b8e344/pkg/runtime/framework/plugins/torch/torch.go#L186-L189

Author: yes. In this case, we don't need this field.

  // NumProcesses is the total number of processes across all hosts.
  // +kubebuilder:validation:Minimum=1
  // +kubebuilder:validation:Required

>   NumProcesses int `json:"numProcesses"`

Reviewer: Is it duplicate to `numNodes` field?

https://github.com/kubeflow/trainer/blob/aca63664ce60226595c83fc50bb8a39b47b8e344/pkg/apis/trainer/v1alpha1/trainingruntime_types.go#L150-L161

  // ProcessID is the unique integer id for this process (0..NumProcesses-1).
  // +kubebuilder:validation:Minimum=0
  // +kubebuilder:validation:Required

>   ProcessID int `json:"processID"`

Reviewer: Can we also generate this?

Author: not sure. i'll check it. and update you.

  // LocalDeviceIDs lists local device indexes assigned to this process (e.g. [0,1]).
  // +kubebuilder:validation:MinItems=1
  // +kubebuilder:validation:Required

>   LocalDeviceIDs []int `json:"localDeviceIDs"`

Reviewer: I think, we can always make the full use of the GPUs/TPUs allocated to the training Pods. Can JAX detect them automatically?

Author: yes. so then if I got you correctly, we shall remove this and let the jax decide.

  // ClusterDetectionMethod (optional) — a hint for automatic detection strategies
  // (e.g., "slurm", "openmpi", "tpu", "env", "none"). If unset and not using an
  // automatic launcher, Distributed must be provided with the required fields above.

>   ClusterDetectionMethod *string `json:"clusterDetectionMethod,omitempty"`

Reviewer: Can you elaborate this field a bit? Is it for initialization? Does it have overlap with the `TargetBackend` field?
